WSLA: Add initial wsladiag tool with basic --list command and placeholder ListSessions

beena352 opened the pull request:
This change adds the first version of the `wsladiag` diagnostic tool and hooks it up to the WSLA host service.

- Adds a new `wsladiag` executable under `src/windows/wsladiag` and integrates it into the CMake build.
- Implements a basic `main.cpp` that parses `--list` and `--help` arguments and prints a placeholder message for `--list`.
- Updates `WSLAUserSession::ListSessions` to return a safe empty result instead of `E_NOTIMPL`, so it is ready to be called from `wsladiag` in a follow-up change.
- Verified that the solution builds and that `wsladiag.exe` runs and accepts `--list` / `--help`.

## Detailed Description of the Pull Request / Additional comments

This PR is the first step toward a WSLA-specific diagnostic tool (`wsladiag`) that can be used by developers to inspect and debug WSLA sessions.

**Tool side (`wsladiag`):**
- Introduces a new `wsladiag` executable under `src/windows/wsladiag`.
- Adds a simple `main.cpp` that:
- Parses `--list` and `--help` command-line options.
- For `--list`, currently prints a placeholder message. A follow-up change will connect this to the WSLA host service `ListSessions` COM API.
- Integrates the new executable into the existing CMake build so it is produced as part of the Windows build.

**Host service side (`WSLAUserSession::ListSessions`):**
- Replaces the `E_NOTIMPL` stub implementation with a safe placeholder that:
- Validates the output parameters.
- Returns an empty `WSLA_SESSION_INFORMATION` list and a `SessionsCount` of `0`.
- This allows callers (such as `wsladiag`) to start wiring up the COM call without hitting `E_NOTIMPL`. A later change will populate the result from `WSLAUserSessionImpl::m_sessions` and return real session information.

This PR is intended as a **starting point / draft** to confirm the overall direction (shape of the tool, build integration, and host API surface) before adding the actual session enumeration and debug-shell functionality.

## Validation Steps Performed

- Built the solution in Debug configuration:
  - `cmake --build . --config Debug`
- Confirmed that `wsladiag.exe` is produced under `bin\x64\Debug`.
- Ran the tool manually:
  - `wsladiag.exe --help` → shows usage text.
  - `wsladiag.exe --list` → prints the placeholder message without errors.
- Verified that updating `WSLAUserSession::ListSessions` from `E_NOTIMPL` to an empty result does not break the service build.

beena352 committed:
WSLA: Add initial wsladiag tool with basic --list support and placeholder ListSessions implementation

OneBlue commented on src/windows/wsladiag/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+set(SOURCES main.cpp)
```
FYI We'll probably want to add this new executable to the MSI package. 

To that, we'll need a add dependency in: https://github.com/microsoft/WSL/blob/feature/wsl-for-apps/msipackage/CMakeLists.txt (so that the MSI package depends on this new `wsladiag` target

And then add an entry in: [package.wix.in](https://github.com/microsoft/WSL/blob/feature/wsl-for-apps/msipackage/package.wix.in) (most likely something similar  to [this](https://github.com/microsoft/WSL/blob/8eee7439b2f1a7bbb0a6d5da04c1bce7b7ec8a0c/msipackage/package.wix.in#L29))

OneBlue commented on src/windows/wsladiag/main.cpp:
```diff
@@ -0,0 +1,44 @@
+
+int wmain(int argc, wchar_t** argv)
+{
+    bool list = false;
```
Since this is a new main() function, we need to do all the initialization like we do for processes. For now we probably to add: 

```
wsl::windows::common::wslutil::ConfigureCrt();
wsl::windows::common::wslutil::InitializeWil();

WslTraceLoggingInitialize(LxssTelemetryProvider, !wsl::shared::OfficialBuild);
auto cleanupTelemetry = wil::scope_exit_log(WI_DIAGNOSTICS_INFO, []() { WslTraceLoggingUninitialize(); });


wsl::windows::common::wslutil::SetCrtEncoding(_O_U8TEXT);

auto coInit = wil::CoInitializeEx(COINIT_MULTITHREADED);
wsl::windows::common::wslutil::CoInitializeSecurity();

WSADATA data;
THROW_IF_WIN32_ERROR(WSAStartup(MAKEWORD(2, 2), &data));

```

A few of those will be required before we can actually call into the COM API

beena352 commented:
@microsoft-github-policy-service agree company="Microsoft"

beena352 committed:
Fix CMakeLists.txt formatting issue for wsladiag

beena352 committed:
Add wsladiag + MSI entries

OneBlue commented on src/windows/wsladiag/main.cpp:
```diff
@@ -0,0 +1,101 @@
+            L"Usage:\n"
+            L"  wsladiag --list    List WSLA sessions\n"
+            L"  wsladiag --help    Show this help",
+            stdout);
```
nit: Other WSL processes currently write help messages to stdout, but since this is a new executable I think we should do the right thing and write message like this on stderr instead

OneBlue commented:
FYI, to format the code to our clang-format rules, you can run `powershell .\FormatSource.ps1 -ModifiedOnly $false`. 

You can also enable clang-format in VS for that to be done automatically.

beena352 committed:
Add initial implementation of --list command to wsladiag

kevpar commented on src/windows/wsladiag/main.cpp:
```diff
@@ -0,0 +1,150 @@
+    wslutil::ConfigureCrt();
+    wslutil::InitializeWil();
+
+    WslTraceLoggingInitialize(LxssTelemetryProvider, !wsl::shared::OfficialBuild);
```
@OneBlue this seems like a strange provider to use here. Is there a more appropriate one?

kevpar replied:
I agree, I think we can just use a single provider for all of WSLA.

beena352 committed:
Update WSLA telemetry provider and improve session enumeration

beena352 committed:
Remove Utf8ToDisplayName and rely on std::formatter<char> for conversion*

OneBlue committed:
wsla: Fix incorrect call to EqualSid causing WSLAUserSession to always be created

beena352 committed:
WIP: before cherry-picking WSLA session fix

beena352 committed:
WSLA: Implement ListSessions and correct CreateSession storage

OneBlue commented on src/windows/wsladiag/CMakeLists.txt:
```diff
@@ -1,3 +1,5 @@
+cmake_minimum_required(VERSION 3.20)  # add this at the very top
```
This shouldn't be needed (it's already in the root CMakeLists.txt)

OneBlue commented on src/windows/wsladiag/CMakeLists.txt:
```diff
@@ -9,6 +11,9 @@ target_link_libraries(wsladiag
 
-target_precompile_headers(wsladiag REUSE_FROM common)
+# Only try to reuse precompiled headers if 'common' exists
+if (TARGET common)
```
`common` always exist, so we can remove the if

OneBlue commented on src/windows/wsladiag/main.cpp:
```diff
@@ -0,0 +1,153 @@
+            wslutil::PrintMessage(std::format(L"Found {} WSLA session{}:\n", sessions.size(), sessions.size() > 1 ? L"s" : L""), stdout);
+
+            wslutil::PrintMessage(L"ID\tCreator PID\tDisplay Name\n", stdout);
+            wslutil::PrintMessage(L"--\t-----------\t------------\n", stdout);
```
This is probably OK as the first implementation, but long term (maybe in a followup PR), we'll want to compute what's the longest string that we'll display, so we can have the right number of columns displayed here so everything is alligned regardless of the DisplayName's sizes

OneBlue replied:
Here's how do it for wsl --list for reference: https://github.com/microsoft/WSL/blob/ba90ee11fa83a1cb7d908d64cb50197afa6a8aeb/src/windows/common/WslClient.cpp#L730C8-L738C12 

(We could probably do a more modern version of that)

beena352 committed:
Merge branch 'feature/wsl-for-apps' into user/beenachauhan/wsladiag

beena352 committed:
Address PR feedback: atomic session ID, CATCH_LOG, fixed-size DisplayName, cleaned wsladiag output

beena352 commented:
Update:
Addressed all review feedback: added std::atomic for session IDs, added try/CATCH_LOG() to ListSessions, switched DisplayName to a fixed-size buffer in the session struct, updated the wsladiag output code to use the new buffer, and applied the CMake/style nits.

beena352 committed:
Address build issues and formatting corrections

beena352 committed:
Sync WSLAVirtualMachine interface changes from upstream

OneBlue commented on src/windows/wslaservice/exe/WSLAUserSession.h:
```diff
@@ -30,14 +32,17 @@ class WSLAUserSessionImpl
     HRESULT CreateSession(const WSLA_SESSION_SETTINGS* Settings, const VIRTUAL_MACHINE_SETTINGS* VmSettings, IWSLASession** WslaSession);
-
+    HRESULT ListSessions(_Out_ WSLA_SESSION_INFORMATION** Sessions, _Out_ ULONG* SessionsCount);
+    void OnVmTerminated(WSLAVirtualMachine* machine);
```
That's probably a merge issue. Let's remove this.
